### PR TITLE
For #46368: Customizable export UI, plus additional hook methods for Shot/Cut updating.

### PR DESCRIPTION
**Pull request by thebeeland**
This is implementing the idea outlined in #41 from CBSD. It's providing hook methods for building custom widgets for each processor we provide, plus allows for custom preset properties to be created, and those properties then being used to set/initialize the custom widget.

In addition, to meet the workflow requirements outlined by CBSD, additional hook methods have been added to allow for customization of if/how Shots, Cuts, and CutItems are created/updated during the export process.

**NOTE:** There are likely going to be use cases for custom UI elements/properties that we then don't have hooks for in the right places to control certain export functionality via those custom properties. The `hiero_update_shot` and `hiero_update_cuts` hooks added here provide the flexibility necessary to implement CBSD's custom shot processor behavior, but doesn't go beyond that.

**Hooks added:**

- hiero_customize_export_ui:
    - create_shot_processor_widget
    - get_shot_processor_ui_properties
    - set_shot_processor_ui_properties
    - create_transcode_exporter_widget
    - get_transcode_exporter_ui_properties
    - set_transcode_exporter_ui_properties
    - create_audio_exporter_widget
    - get_audio_exporter_ui_properties
    - set_audio_exporter_ui_properties
    - create_nuke_shot_exporter_widget
    - get_nuke_shot_exporter_ui_properties
    - set_nuke_shot_exporter_ui_properties
- hiero_update_shot:
    - create_filesystem_structure
    - update_shotgun_shot_entity
- hiero_update_cuts:
    - allow_cut_updates
    - create_cut_item
    - get_cut_thumbnail

**Example custom UI based on CBSD's example:**

image

### CODE REVIEW

**Comment by dashielb1010** [comment]
Wow this is great!
So cool that these hooks are becoming a reality. I'm very happy you've taken the time to incorporate them. Looking forward to testing them out when my schedule frees up! -Cheers!

File: info.yml [hunk 1]
```diff
@@ -172,6 +172,29 @@ configuration:
...
+        type: hook
+        description: "Called during SG shot processing. Provides methods for managing
+                     if and how the Shot's entity in updated, and if and how creation
+                     of the Shots filesystem structure happens during export."
+        parameters: []
+        default_value: hiero_update_shot
+
+    hook_update_cuts:
+        type: hook
+        description: ""
```
Comment (thebeeland) [inline info.yml]: Whoops. Missed this one. I'll get a description in!

**Comment by manneohrstrom** [comment]
Adding josh as well - his input will be very valuable here given that he has been working extensively both with the publish2 UI extensions and the hiero exporter.

File: hooks/hiero_customize_export_ui.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+        :param widget: The Qt widget that was created by the associated create
+            widget hook method.
+        :param OrderedDict properties: A dict containing property widget
+            objects, keyed by label, that were constructed from the data
+            built by the associated get properties hook method.
+        """
+        return
+
+    def create_audio_exporter_widget(self, parent):
```
Comment (manneohrstrom) [inline hooks/hiero_customize_export_ui.py]: if these methods all follow the same pattern, could we have a single set of three methods where we also pass a 'type' parameter? I guess this depends on the patterns we envision people adopting - are we thinking users may override these hooks piecemeal, e.g. say just the experter widget? Or all combined? If the latter, then the type approach might work well and be a lot simpler? If the former, it makes sense to keep them separated i guess?

File: hooks/hiero_customize_export_ui.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+from sgtk.platform.qt import QtGui
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class HieroCustomizeExportUI(HookBaseClass):
+    """
+    This class defines methods that can be used to customize the UI of the various
+    Shotgun-related exporters. Each processor has its own set of create/get/set
```
Comment (manneohrstrom) [inline hooks/hiero_customize_export_ui.py]: the separate examples snippets are great but i think adding an full example would be really useful too. Maybe even something that could be cut n pasted into a hook file as a starting point for TDs!

File: hooks/hiero_customize_export_ui.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            ]
+
+        Example Implementation:
+
+        .. code-block:: python
+
+            return [
+                dict(
+                    label="Create Cut:",
+                    key="sgCreateCut",
```
Comment (manneohrstrom) [inline hooks/hiero_customize_export_ui.py]: what is this `sgCreateCut` - what does it signify?

File: hooks/hiero_update_cuts.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            pairs to use when creating the CutItem entity in Shotgun.
+        :param dict preset_properties: The export preset's properties
+            dictionary.
+
+        :returns: The created CutItem entity dictionary, or None if
+            no CutItem entity was created.
+        :rtype: dict or None
+        """
+        cut_item = self.parent.sgtk.shotgun.create("CutItem", cut_item_data)
+        self.parent.logger.info("Created CutItem in Shotgun: %s" % cut_item)
```
Comment (manneohrstrom) [inline hooks/hiero_update_cuts.py]: will the fact that we don't batch these guys really slow things down?
Could the hook take a list of cut items and do a batch?

File: hooks/hiero_update_cuts.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            is to be uploaded to Shotgun.
+        :rtype: str or None
+        """
+        thumbnail = None
+        hiero_sequence = task_item.sequence()
+
+        try:
+            # See if we can find a poster frame for the sequence and
+            # turn that into a usable thumbnail.
+            thumbnail = hiero_sequence.thumbnail(hiero_sequence.posterFrame())
```
Comment (manneohrstrom) [inline hooks/hiero_update_cuts.py]: can we provide a link to a hiero API reference here?

File: hooks/hiero_update_shot.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    This class defines methods that handle updating the Shot entity
+    in Shotgun, as well as whether and how the filesystem structure
+    is created for a Shot during export.
+    """
+    def create_filesystem_structure(self, entity_type, entity_id, preset_properties):
+        """
+        Handles creating the filesystem structure for the shot that
+        was exported. The preset properties dictionary is provided to
+        allow for the lookup of any custom properties that might have
+        been defined in other hooks, like can be achieved when using
```
Comment (manneohrstrom) [inline hooks/hiero_update_shot.py]: a practical example to illustrate here would be great.

File: hooks/hiero_update_shot.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            "Shot".
+        :param int entity_id: The id of the entity that was created
+            or updated as part of the export.
+        :param dict preset_properties: The export preset's properties
+            dictionary.
+        """
+        self.parent.logger.debug(
+            "Creating file system structure for %s %s..." % (entity_type, entity_id)
+        )
+        self.parent.sgtk.create_filesystem_structure(entity_type, [entity_id])
```
Comment (manneohrstrom) [inline hooks/hiero_update_shot.py]: keep in mind that the `self.parent` in hooks will require a fairly modern core.

File: hooks/hiero_customize_export_ui.py (hunk 4 above)
Comment (thebeeland) [reply hooks/hiero_customize_export_ui.py]: I used the kwarg name from Hiero's ui property constructor, but maybe we can make it a little more obvious by calling it "property_name" or "name" instead. Label is the human-readable name that's displayed in the UI, and key is the "system" name of the property.

File: hooks/hiero_update_cuts.py (hunk 5 above)
Comment (thebeeland) [reply hooks/hiero_update_cuts.py]: I can look at the underlying logic and see what might be possible. Is it too slow now? This isn't changing the logic flow other than running it from a hook.

File: hooks/hiero_update_shot.py (hunk 8 above)
Comment (thebeeland) [reply hooks/hiero_update_shot.py]: tk-nuke already requires v0.18.45

File: hooks/hiero_customize_export_ui.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+                    value=True,
+                    type=bool,
```
[the remark this answers is not in the document]
Comment (thebeeland) [reply hooks/hiero_customize_export_ui.py]: Yeah...I just tried to think of an example of where it might be different than type(value), and I can't come up with anything reasonable. I'll ditch it!

File: hooks/hiero_update_shot.py (hunk 8 above)
Comment (thebeeland) [reply hooks/hiero_update_shot.py]: And looking deeper, tk-hiero-export requires v0.17.21 of tk-core, which looks like it contains the parent property for Hook.

File: hooks/hiero_customize_export_ui.py (hunk 4 above)
Comment (thebeeland) [reply hooks/hiero_customize_export_ui.py]: Changed it to "name"

Comment (thebeeland) [reply hooks/hiero_customize_export_ui.py]: I'll also change the example to use a more verbose name

File: hooks/hiero_customize_export_ui.py [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+            widget.setLayout(QtGui.QFormLayout())
+            return widget
+
+        :param parent_widget: The parent widget.
+
+        :returns: A custom widget.
+        """
+        return None
+
+    def get_shot_processor_ui_properties(self):
```
[the remark this answers is not in the document]
Comment (thebeeland) [reply hooks/hiero_customize_export_ui.py]: Hmm...I was just doing this, and it feels odd. Everything related to this on the Hiero/NS end of things revolves around "properties" and not "settings." It's feeling really strange mixing the two nomenclatures interchangeably. Having the getter return "settings dictionaries" feels okay, generally speaking, but then having the setter provide a list of property objects from Hiero's API and calling those "settings" doesn't feel right.

Thoughts?

**Comment by thebeeland** [comment]
I've added a first pass of the sphinx docs setup. As noted in the commit message, there are a couple caveats when compared to publish2 and tk-nuke-quickreview:

- The execute methods in old-style hooks have to pass instead of raising NotImplementedError.
- The base_hooks can't live under the tk_hiero_export module, because importing from tk_hiero_export requires access to the hiero api, which isn't available at docs build time.
- The docs for most of the older are VERY minimal, so it doesn't look great. We should probably include them anyway, though, and queue up a ticket for a future sprint to do a docs pass on those older hooks.

File: hooks/hiero_customize_export_ui.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+# ===========================
+
+import sgtk
+
+from sgtk.platform.qt import QtGui
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class HieroCustomizeExportUI(HookBaseClass):
```
[the remark this answers is not in the document]
Comment (thebeeland) [reply hooks/hiero_customize_export_ui.py]: First pass done. See the comment below for more detail on a few caveats.

File: hooks/hiero_update_cuts.py (hunk 6 above)
Comment (thebeeland) [reply hooks/hiero_update_cuts.py]: Done.

File: hooks/hiero_update_shot.py (hunk 7 above)
Comment (thebeeland) [reply hooks/hiero_update_shot.py]: Done.

File: hooks/hiero_customize_export_ui.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+    """
+    This class defines methods that can be used to customize the UI of the various
+    Shotgun-related exporters. Each processor has its own set of create/get/set
+    methods, allowing for customizable UI elements for each type of export.
+    """
+    def create_shot_processor_widget(self, parent_widget):
+        """
+        Builds and returns a custom widget to be embedded in the parent exporter.
```
Comment (josh-t) [inline hooks/hiero_customize_export_ui.py]: Worth mentioning how it will fit into the layout?

File: hooks/hiero_customize_export_ui.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (josh-t) [reply hooks/hiero_customize_export_ui.py]: I'd say stick with Hiero/NS terms

File: hooks/hiero_customize_export_ui.py (hunk 2 above)
Comment (josh-t) [reply hooks/hiero_customize_export_ui.py]: I was thinking about a `type` argument as well, but I think you'd just end up with a big conditional that called out to other methods in the hook.

One idea that could simplify things a bunch would be to define a base hook that only has 3 methods (create widget and get/set properties). This could be used for any of the types, but you associate the type/hook at the config level. So if you need a custom audio and transcode hooks, you'd configure something like this:
```yaml
custom_export_ui_plugins:
  - type: audio
    hook: "{config}/hooks/tk-hiero-export/custom_audio.py"
  - type: transcode
    hook: "{config}/hooks/tk-hiero-export/custom_transcode.py"
```
Now you can cut down on the boilerplate and just define the API that's used for all the various types. Would that work?

File: hooks/hiero_update_shot.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    def create_filesystem_structure(self, entity_type, entity_id, preset_properties):
+        """
+        Handles creating the filesystem structure for the shot that
+        was exported. The preset properties dictionary is provided to
+        allow for the lookup of any custom properties that might have
+        been defined in other hooks, like can be achieved when using
+        the hiero_customize_export_ui hook.
+
+        :param str entity_type: The entity type that was created or
```
Comment (josh-t) [inline hooks/hiero_update_shot.py]: If we're breaking this out into a hook, I'm wondering if we should deemphasize "Shot" to some extent. Maybe something as simple as renaming the hook/class/methods to not include the word "shot".

File: hooks/hiero_update_shot.py (hunk 8 above)
Comment (josh-t) [reply hooks/hiero_update_shot.py]: If we're doing the hook docs and base hooks, it's going to need a really new core anyway. :)

File: python/tk_hiero_export/shot_updater.py [hunk 14]
```diff
@@ -254,41 +260,61 @@ def taskStep(self):
             sg_shot['task_template'] = template
 
         # commit the changes and update the thumbnail
-        self.app.log_debug("Updating info for %s %s: %s" % (shot_type, shot_id, str(sg_shot)))
-        self.app.tank.shotgun.update(shot_type, shot_id, sg_shot)
+        self.app.execute_hook_method(
```
Comment (josh-t) [inline python/tk_hiero_export/shot_updater.py]: 👍

File: python/tk_hiero_export/shot_updater.py [hunk 15]
```diff
@@ -254,41 +260,61 @@ def taskStep(self):
...
+            entity_type=shot_type,
+            entity_id=shot_id,
+            entity_data=sg_shot,
+            preset_properties=self._preset.properties(),
+        )
 
         # create the directory structure
-        self.app.log_debug("Creating file system structure for %s %s..." % (shot_type, shot_id))
-        self.app.tank.create_filesystem_structure(shot_type, [shot_id])
+        self.app.execute_hook_method(
```
Comment (josh-t) [inline python/tk_hiero_export/shot_updater.py]: 👍

**Comment by josh-t** [comment]
One more thing: You might consider creating hook instances throughout the app and keeping them persistent rather than the old way. That way you can instantiate them with the new base classes to get any default implementation, allowing clients to only override methods they're customizing.

**Comment by josh-t** [comment]
Another though... sorry...

Would be nice to have some sample/test implementation (looks like you've already written) that we could use for QA. The publish2 app does something similar ([see here](https://github.com/shotgunsoftware/tk-multi-publish2/tree/master/tests)).

**Comment by manneohrstrom** [comment]
Thanks for the sphinx docs pass 1! Awesome!!! Josh made some great points about the overall flow, so I'll hold off further changes for now until you have reviewed that feedback or i might confuse things :)

About the old methods vs sphinx - definitely huge value to get them into sphinx even if they count have fully featured docs like the new ones - i think it would be great if we could at least get basic params and descriptions in as part of this PR - hopefully that won't take a huge amount of time? 

This is very exciting!

File: hooks/hiero_update_shot.py (hunk 8 above)
Comment (thebeeland) [reply hooks/hiero_update_shot.py]: What's the minimum required for that? I assume it's newer than 0.17.21, which means I should bump up the requirement in info.yml.

File: hooks/hiero_customize_export_ui.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (thebeeland) [reply hooks/hiero_customize_export_ui.py]: I'm leaning that way, as well.

File: hooks/hiero_update_shot.py (hunk 13 above)
Comment (thebeeland) [reply hooks/hiero_update_shot.py]: I'm leaning towards not changing this from "shot", simply because it's more or less paired with the existing hiero_get_shot hook. We can't easily rename that hook without potentially breaking client setups that make use of that hook, so I'm inclined to keep get/update shot hooks sharing nomenclature.

**Comment by manneohrstrom** [comment]
One more thought - for the next round of review would be great if you could zip up and attach a sphinx doc build to the PR! Just in case it's tricky to build locally. (e.g. code imports hiero stuff etc).

File: python/tk_hiero_export/base.py [hunk 16]
```diff
@@ -34,6 +36,84 @@ def setApp(cls, app):
...
+        if get_method not in self._custom_property_definitions:
+            self._custom_property_definitions[get_method] = self.app.execute_hook_method(
+                "hook_customize_export_ui",
+                get_method,
+            )
+
+        return self._custom_property_definitions[get_method]
+
+    def _get_custom_widget(self, parent, create_method, get_method, set_method, properties=None):
+        """
```
[the remark this answers is not in the document]
Comment (thebeeland) [reply python/tk_hiero_export/base.py]: Fixed

**Comment by thebeeland** [comment]
After discussing it with Josh and Rob, we're not going to go the full publish2-style setup for how the hooks are configured. Instead, we're keeping the single hook for the custom export ui, but cleaning it up a bit. I've gone through and set base classes pointing back to the base_hooks (even for the old hooks), and I've put the no-op implementation of each for the create/get/set hook methods for custom ui into the base hook. I've then emptied out the hooks/hiero_customize_export_ui.py file, and linked to the (eventual) sphinx docs page in a comment. @manneohrstrom -- If you have any better suggestions around that last bit, please let me know!

I've also added some basic docs to the older hooks, and things look much better in sphinx docs as a result.

Back to you guys! @josh-t @manneohrstrom

File: hooks/hiero_update_shot.py (hunk 8 above)
Comment (josh-t) [reply hooks/hiero_update_shot.py]: I believe 2 releases ago `v0.18.124`

**Comment by manneohrstrom** [comment]
For reference, here are the built docs:

[docs.zip](https://github.com/shotgunsoftware/tk-hiero-export/files/1780764/docs.zip)

File: docs/index.rst [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+- Status of associated Shot entities can be updated
+- Nuke scripts can be written into the project's filesystem structure for each Shot that's processed
+- Sequence and Shot entity data can be updated in Shotgun
+- Cuts can be update to include new CutItems built from the exported sequences
+
+Several aspects of the process are customizable via the following Hooks:
+
+.. py:currentmodule:: base_hooks
+
+Creating custom UI elements
```
Comment (manneohrstrom) [inline docs/index.rst]: I think it's better that we order the hooks according to frequency - start with the ones that are most likely to be used by users, and the most advanced ones at the end!

File: docs/index.rst [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+###################################################
+
+
+The ``tk-hiero-export`` app adds custom Shotgun export processors to the Hiero/Nuke Studio export framework.
+
+Via custom export processors, the user can create and update entities in the current Project in Shotgun.
+
+During the export process a number things can happen:
+
+- Status of associated Shot entities can be updated
```
Comment (manneohrstrom) [inline docs/index.rst]: i would ❤️ if there was a flow chart diagram showing the process and included all the hooks. I think that would be a perfect overview for TDs to understand what is possible. happy to help putting this together if you need help.

File: docs/index.rst [hunk 19]
```diff
@@ -0,0 +1,106 @@
+Shotgun Hiero Export API reference, |release|
```
Comment (manneohrstrom) [inline docs/index.rst]: There is something weird about the doc build - not sure where to put the comment so adding it here - the TOC seems to contain the same items twice. I just realised that we have the same issue for the [nuke review](http://developer.shotgunsoftware.com/tk-nuke-quickreview/) docs but not for the [publish](http://developer.shotgunsoftware.com/tk-multi-publish2/). Could you investigate and potentially update the nuke review docs too?

File: docs/index.rst [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+Via custom export processors, the user can create and update entities in the current Project in Shotgun.
+
+During the export process a number things can happen:
+
+- Status of associated Shot entities can be updated
+- Nuke scripts can be written into the project's filesystem structure for each Shot that's processed
+- Sequence and Shot entity data can be updated in Shotgun
+- Cuts can be update to include new CutItems built from the exported sequences
+
```
Comment (manneohrstrom) [inline docs/index.rst]: Is there good hiero API documentation? Perhaps it's worth including a link here in the intro?

File: app.py [hunk 21]
```diff
@@ -50,6 +50,9 @@
     ShotgunAudioExporterUI,
     ShotgunHieroObjectBase,
 )
+
+import base_hooks
```
Comment (manneohrstrom) [inline app.py]: Why the absolute import? What if this app runs next another app which also imports base hooks? 

We typically very very rarely do absolute imports and if we do, we name things explicitly (e.g. the `import tk_hiero_export` above is slightly better but still unconventional!)

See the nuke review app for another way to do this: https://github.com/shotgunsoftware/tk-nuke-quickreview/blob/master/app.py#L30

File: app.py [hunk 22]
```diff
@@ -61,11 +64,20 @@
 
 
 class HieroExport(Application):
+    _BASE_HOOKS = base_hooks
```
Comment (manneohrstrom) [inline app.py]: missing docstring. Please comment what this constant is for. (why is it needed?)

File: hooks/hiero_customize_export_ui.py [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class HieroCustomizeExportUI(HookBaseClass):
+    """
+    This class defines methods that can be used to customize the UI of the various
+    Shotgun-related exporters. Each processor has its own set of create/get/set
+    methods, allowing for customizable UI elements for each type of export.
+    """
+    # For detailed documentation of the methods available for this hook, see
+    # the documentation at http://developer.shotgunsoftware.com/tk-hiero-export/
```
Comment (manneohrstrom) [inline hooks/hiero_customize_export_ui.py]: nice!

File: hooks/hiero_get_extra_publish_data.py [hunk 24]
```diff
@@ -12,14 +12,23 @@
...
     def execute(self, task, **kwargs):
         """
-        :param task: The Hiero Task that is currently being processed
-        Return the dictionary to update the data for the PublishedFile in Shotgun
-        or None if there is no extra information to publish.
+        Get a data dictionary for a PublishedFile to be updated in Shotgun.
 
-        The track item associated with this task can be accessed via task._item.
+        .. note:: The track item associated with this task can be accessed via
+            task._item.
```
Comment (manneohrstrom) [inline hooks/hiero_get_extra_publish_data.py]: Why are we recommending users to access private members? (sorry if this is out of scope of this PR...)

File: hooks/hiero_pre_export.py [hunk 25]
```diff
@@ -13,13 +13,14 @@
...
     """
     def execute(self, processor, **kwargs):
         """
-        Allows clearing of caches prior to shot processing. This is called just prior to export.
+        Called just prior to export. One use case for would be to clear
+        cached data here, just before the export begins.
 
-        :param processor: Processor The is being used, in case distinguishing between
-                          differnt exports is needed.
+        :param processor: The processor object that is about to be started.
```
Comment (manneohrstrom) [inline hooks/hiero_pre_export.py]: what is a processor? Can we link to hiero docs?

File: hooks/hiero_update_cuts.py (hunk 5 above)
Comment (manneohrstrom) [reply hooks/hiero_update_cuts.py]: is this new or old code? I am not sure about current performance. But if you are running say a 200 shot conform, running create 200 times will definitely be slower than running a batch. But like you say, it depends on other factors, workflow ux etc. If this is old code and we are merely updating the docs, probably better to leave things the way they are!

File: hooks/hiero_update_cuts.py (hunk 6 above)
Comment (manneohrstrom) [reply hooks/hiero_update_cuts.py]: why not do it as part of the sphinx docs?

File: python/base_hooks/hiero_customize_export_ui.py [hunk 26]
```diff
@@ -0,0 +1,257 @@
...
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+#  UI Hook
+# ===========================
```
Comment (manneohrstrom) [inline python/base_hooks/hiero_customize_export_ui.py]: nitpick: maybe drop this header. We don't normally have these kinds of things.

File: python/base_hooks/hiero_customize_export_ui.py [hunk 27]
```diff
@@ -0,0 +1,257 @@
...
+
+import sgtk
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class HieroCustomizeExportUI(HookBaseClass):
+    """
+    This class defines methods that can be used to customize the UI of the various
```
Comment (manneohrstrom) [inline python/base_hooks/hiero_customize_export_ui.py]: could we include a screenshot to illustrate as part of the docs?

File: hooks/hiero_update_cuts.py (hunk 6 above)
Comment (thebeeland) [reply hooks/hiero_update_cuts.py]: Can do!

File: docs/index.rst (hunk 19 above)
Comment (thebeeland) [reply docs/index.rst]: I'll ping you offline about this.

File: docs/index.rst (hunk 18 above)
Comment (thebeeland) [reply docs/index.rst]: Agreed, we can chat offline about this.

File: python/base_hooks/hiero_customize_export_ui.py (hunk 27 above)
Comment (thebeeland) [reply python/base_hooks/hiero_customize_export_ui.py]: Done. Also added some overview sphinx docs for customizing UI. Basic stuff, but it's a good, brief intro with an example screenshot.

File: docs/index.rst (hunk 19 above)
Comment (thebeeland) [reply docs/index.rst]: We have an internal ticket to look at this more closely. It's not immediately obvious why the ToC is wonky.
